zcash_transparent/src/zip48.rs

```diff
@@ -0,0 +1,450 @@
+    ///
+    /// [BIP 388 `KEY_INFO` expression]: https://github.com/bitcoin/bips/blob/master/bip-0388.mediawiki#key-information-vector
+    pub fn key_info_expression<P: consensus::Parameters>(&self, params: &P) -> String {
+        self.key_expression_inner(pub_prefix(params), vec![])
+            .to_string()
+    }
+
+    /// Produces a [`KeyExpression`] from this public key corresponding to a specific
+    /// address.
+    pub fn key_expression_for_address(
+        &self,
+        prefix: Prefix,
+        scope: zip32::Scope,
+        address_index: NonHardenedChildIndex,
+    ) -> KeyExpression {
+        self.key_expression_inner(
+            prefix,
+            vec![
+                TransparentKeyScope::from(scope).into(),
+                address_index.into(),
+            ],
+        )
+    }
+
+    fn key_expression_inner(
+        &self,
+        prefix: Prefix,
+        child: Vec<bip32::ChildNumber>,
+    ) -> KeyExpression {
+        KeyExpression::from_xpub(Some(self.origin.clone()), prefix, self.key.clone(), child)
+            .expect("correct by construction")
+    }
+}
+
+/// A [ZIP 48] P2SH multisig full viewing key.
+///
+/// This provides the necessary derivation capability to view all funds controlled by a
+/// P2SH multisig account.
+///
+/// [ZIP 48]: https://zips.z.cash/zip-0048
+pub struct FullViewingKey {
+    threshold: u8,
+    key_info: Vec<AccountPubKey>,
+}
+
+impl FullViewingKey {
+    /// Constructs a full viewing key for a standard ZIP 48 account.
+    ///
+    /// This uses the standard wallet descriptor template. For example:
+    /// - `threshold == 2`
+    /// - `key_info.len() == 3`
+    /// - Wallet descriptor template: `"sh(sortedmulti(2,@0/**,@1/**,@2/**))"`
+    pub fn standard(
+        threshold: u8,
+        key_info: Vec<AccountPubKey>,
+    ) -> Result<Self, FullViewingKeyError> {
+        if key_info.is_empty() {
+            Err(FullViewingKeyError::NoPubKeys)
+        } else if key_info.len() > 15 {
+            Err(FullViewingKeyError::TooManyPubKeys)
+        } else if usize::from(threshold) > key_info.len() {
+            Err(FullViewingKeyError::InvalidThreshold)
+        } else {
+            // Verify `key_info` in a scope so we can borrow from it.
+            {
+                // To be compatible with ZIP 48, all keys must have the same derivation
+                // information. We check this by collecting into a set, and then checking
+                // it contains one entry. We don't need to verify the derivation's
+                // structure because `AccountPubKey` enforces it by construction.
+                let derivations = key_info
+                    .iter()
+                    .map(|key| key.origin.derivation())
+                    .collect::<BTreeSet<_>>();
+                if derivations.len() != 1 {
+                    return Err(FullViewingKeyError::IncompatiblePubKeys);
+                }
```

Comment:
Notes:
- This requires BIP 380 key origin information to be provided by all participants. It's optional in BIP 380 but I think fine for us to require in ZIP 48.
- This currently requires the `account` value to be identical for all keys in a ZIP 48 account. BIP 48 is very underspecified but every example of its use that I can find uses the same account number for each participant, and it's easier to lift this restriction in future than to impose it later, so for now I'm imposing it.

I plan to open a PR with these updates to the ZIP 48 draft.